.github/pull_request_template.md

```diff
@@ -0,0 +1,53 @@
+Fixes #XXXX.
+
+(Replace XXXX with the id of the issue fixed in this PR. Remove this line if there is no corresponding
+issue. Don't reference the issue in the title of this pull-request.)
+
+Add tags to your PR if you are a committer (only committers have the right to add tags). Add [Design Review] tag
+if this PR should better be reviewed by at least two people.
+Don't forget to add the following tags (if applicable): [Incompatible], [Release Notes], [Compatibility], [Security],
+[Development Blocker]. Add at least one [Area - ] tag, consider creating a new one if none of the existing [Area - ]
+tags is applicable.
+
+### Description
+
+Describe the goal of this PR, what problem are you fixing. If there is a corresponding issue (referenced above), it's
+not necessary to repeat the description here, however, you may choose to keep one summary sentence.
+
+Describe your patch: what did you change in code? How did you fix the problem?
+
+If there are several relatively logically separate changes in this PR, list them. For example:
+ - Fixed the bug ...
+ - Renamed the class ...
+ - Added a forbidden-apis entry ...
+
+Some of the aspects mentioned above may be omitted for simple and small PRs.
+
+### Design
+
+Please describe any design decisions made, including:
+ - Choice of algorithms
+ - Behavioral aspects. What configuration values are acceptable? How are corner cases and error conditions handled, such
+   as when insufficient resources are available?
+ - Class organization and design (how the logic is split between classes, inheritance, composition, design patterns)
+ - Method organization and design (how the logic is split between methods, parameters and return types)
+ - Naming (class, method, API, configuration, HTTP endpoint, names of emitted metrics)
+
+In addition, describe _at least one_ alternative design (or mention alternative name) for every design (or naming)
```

Comment:
This is similar to the "discuss other alternative solutions" guidance in the proposal template. However, the proposal template doesn't require it. I don't see a reason to require it. It might end up being artificial if forced.

Reply:
@gianm A section "Design It Twice" in [this book](https://www.amazon.com/Philosophy-Software-Design-John-Ousterhout/dp/1732102201) discusses this aspect and explains the benefits of trying to come up with an alternative, regardless of how obvious and undebatable the implemented design seems to be and how ridiculous the alternatives seem to be. It says that this excersize often helps to uncover important properties of the design and come up with a third design, better than the original one and the alternative.

Reply:
I guess I can't speak to the argument in that book because I don't have a copy of it. But even though there are benefits to the thought experiment of thinking through alternate designs, I don't think we need to make it a hard requirement that people do it for every facet of every contribution. How about converting this to a suggestion rather than a requirement.

Reply:
I suspect that even with this template, in some PRs design won't be discussed at all, because the first action of the contributor after opening the "New PR" interface on Github will be selecting all the template text and deleting it. People who contribute regularly will do the same (and that's ok) because they already read this template text once.

I don't worry that anybody will ever take this instruction too far. I'm sure that even with the current wording, contributors will still under-discuss the design rather than over-discuss it, on average. I think that softer wording will shift this even more towards under-discussion.

Reply:
I don't think writing intentionally over-strict requirements is the right way to encourage good behavior. I think we should show a little more trust in contributors and write requirements that reflect what we actually expect from them (and from each other, since we are all contributors). It's certainly good to list out suggestions / things to consider that make a PR better, but phrasing them as requirements seems unwelcoming and disingenuous to my ears.

Reply:
Always considering alternative designs is a good ideal, but in practice, If we had a hard requirement for the "describe alternatives" rule, I would expect many PRs to have empty fluff in that section or people just describing some awful alternative implementation to check off the requirement, I expect it would often not bring any real value if forced. 

If the PR submitter shares in the spirit behind this rule, they would likely do if it's optional. If they don't, they'll ignore it or game the system.

If I saw an otherwise good PR that lacked this section, I would personally not be willing to reject it on that basis alone.

So I would agree with making this optional.

Reply:
I also think making it a hard requirement shifts too much burden to the PR author.

If there are interesting alternative designs, a reviewer can bring them up during review.

If it's a PR where no reviewer can think of alternative designs that are worth considering, then I don't think it's reasonable to place that requirement on the PR author either.

Reply:
I'll update to more optional language because I don't see a point in fighting for this.

Reply:
Just answering:

> I expect it would often not bring any real value if forced.

That's an intuitive expectation, but the counterintuitive point is that it actually brings real value far more often then expected.

> If I saw an otherwise good PR that lacked this section, I would personally not be willing to reject it on that basis alone.

I don't any think problems with PR decoration is a basis for regecting any PR at all.

> I also think making it a hard requirement shifts too much burden to the PR author.
> 
> If there are interesting alternative designs, a reviewer can bring them up during review.
> 
> If it's a PR where no reviewer can think of alternative designs that are worth considering, then I don't think it's reasonable to place that requirement on the PR author either.

Authors are in a far better position to propose alternative designs. Reviewers often barely grasp the idea of the PR, and they only review the code on Github. Authors wrote the code, worked in IDE, understand all ins and outs.

In general, I think it's a good idea for PR authors to place as much burden as they can on themselves. Self-reviewing touches the same idea. Reasons:
 - Self-reliance and removing burden from colleagues is always a good idea. (Doing as much work as we can ourselves, collaborators will care of their work themselves. Not relying on collaborators or quality of their work.)
 - Perhaps when we review our own code (and design decisions) we more likely learn something and improve our own style than when we review other people's code. The same phycological effect as behind "People don't learn from other's people mistakes, only their own." Or maybe because when we ourselves the source of "review comments" we don't tend to subconsciously reject them or treat them passive-aggressively.

But that's a whole other topic of discussion and I'm not going to mention anything of that in the template.

Reply:
> That's an intuitive expectation, but the counterintuitive point is that it actually brings real value far more often than expected.

That's fair, I don't think I would really know the effects of such a rule unless put in practice. But I think doing so is risky. 

Some PRs are submitted by people with a pressing need to have that feature in Druid (e.g., their business relies on a certain feature they made), I don't expect complex PR submission requirements to hinder such people.

But as the PR submissions become more "casual", I think the likelihood of people being discouraged by complex requirements increases. I would personally rather put more burden on reviewers for ensuring quality than potentially hindering incoming contributions for this specific point.

> I don't think any problem with PR decoration is a basis for rejecting any PR at all.

Hm, if I compare description of alternative designs to a description of the chosen design, I would be quite willing to hold off review on a PR or reject it until the author provides a description of what the PR is doing.

To the earlier point about enforcing the rule, I think it's true that making it a "hard requirement" could be psychologically useful, but that too I think has costs. As a general principle, I believe that having rules that aren't enforced hurts the credibility of the body of rules as a whole and introduces some uncertainty.